Fire spreading to invalid locations crashes the server

<!-- Before reporting a crash here, please make sure you are on the latest version of Paper.  -->
<!-- Forks of Paper receive no support here. Please make sure that this crash also happens when using Paper.  -->

### Link/paste of stack trace

```
[20:39:18 ERROR]: Encountered an unexpected exception
net.minecraft.server.v1_16_R2.ReportedException: Exception while ticking
	at net.minecraft.server.v1_16_R2.MinecraftServer.b(MinecraftServer.java:1373) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.DedicatedServer.b(DedicatedServer.java:371) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.MinecraftServer.a(MinecraftServer.java:1211) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.MinecraftServer.w(MinecraftServer.java:999) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.MinecraftServer.lambda$a$0(MinecraftServer.java:177) ~[patched_1.16.3.jar:git-Paper-202]
	at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 16
	at net.minecraft.server.v1_16_R2.PlayerChunk.a(PlayerChunk.java:368) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.ChunkProviderServer.flagDirty(ChunkProviderServer.java:905) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.WorldServer.notify(WorldServer.java:1570) ~[patched_1.16.3.jar:git-Paper-202]
	at org.bukkit.craftbukkit.v1_16_R2.block.CraftBlockState.update(CraftBlockState.java:185) ~[patched_1.16.3.jar:git-Paper-202]
	at org.bukkit.craftbukkit.v1_16_R2.block.CraftBlockState.update(CraftBlockState.java:167) ~[patched_1.16.3.jar:git-Paper-202]
	at org.bukkit.craftbukkit.v1_16_R2.event.CraftEventFactory.handleBlockSpreadEvent(CraftEventFactory.java:787) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.BlockFire.tickAlways(BlockFire.java:219) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.BlockBase$BlockData.a(BlockBase.java:558) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.WorldServer.b(WorldServer.java:803) ~[patched_1.16.3.jar:git-Paper-202]
	at com.destroystokyo.paper.server.ticklist.PaperTickList.tick(PaperTickList.java:306) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.TickListServer.b(TickListServer.java:47) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.WorldServer.doTick(WorldServer.java:461) ~[patched_1.16.3.jar:git-Paper-202]
	at net.minecraft.server.v1_16_R2.MinecraftServer.b(MinecraftServer.java:1358) ~[patched_1.16.3.jar:git-Paper-202]
	... 5 more
```

### Plugin list:
None required; I used WorldEdit to make this easy to reproduce, however.

### Actions to reproduce (if known):

1. Lay out a large square of leaves at Y 0.
1. `//overlay fire`
1. Observe the server crash

### Paper version:

`This server is running Paper version git-Paper-202 (MC: 1.16.3) (Implementing API version 1.16.3-R0.1-SNAPSHOT)`

--- a/Spigot-Server-Patches/0090-Prevent-Fire-from-loading-chunks-wrongly-spread.patch
+++ b/Spigot-Server-Patches/0090-Prevent-Fire-from-loading-chunks-wrongly-spread.patch
@@ -1,13 +1,15 @@
 From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
 From: Aikar <aikar@aikar.co>
 Date: Sun, 17 Apr 2016 17:27:09 -0400
-Subject: [PATCH] Prevent Fire from loading chunks
+Subject: [PATCH] Prevent Fire from loading chunks & wrongly spread
 
 This causes the nether to spam unload/reload chunks, plus overall
 bad behavior.
 
+This also stops fire from spreading to illegal locations.
+
 diff --git a/src/main/java/net/minecraft/server/BlockFire.java b/src/main/java/net/minecraft/server/BlockFire.java
-index 3d6aae40cab0cefbfc56389d9d2d37d2f0badd98..48b52b58ff4779fe92d17ab3b3c4cf608917a325 100644
+index 3d6aae40cab0cefbfc56389d9d2d37d2f0badd98..e79f3872de2989d0b9b28a8b49cc399968f5bdbf 100644
 --- a/src/main/java/net/minecraft/server/BlockFire.java
 +++ b/src/main/java/net/minecraft/server/BlockFire.java
 @@ -115,7 +115,7 @@ public class BlockFire extends BlockFireAbstract {
@@ -23,7 +25,7 @@ index 3d6aae40cab0cefbfc56389d9d2d37d2f0badd98..48b52b58ff4779fe92d17ab3b3c4cf60
                                  }
  
                                  blockposition_mutableblockposition.a((BaseBlockPosition) blockposition, l, j1, i1);
-+                                if (!worldserver.isLoaded(blockposition_mutableblockposition)) continue; // Paper
++                                if (blockposition_mutableblockposition.isInvalidYLocation() || !worldserver.isLoaded(blockposition_mutableblockposition)) continue; // Paper
                                  int l1 = this.a((IWorldReader) worldserver, (BlockPosition) blockposition_mutableblockposition);
  
                                  if (l1 > 0) {

--- a/Spigot-Server-Patches/0258-Add-TNTPrimeEvent.patch
+++ b/Spigot-Server-Patches/0258-Add-TNTPrimeEvent.patch
@@ -5,7 +5,7 @@ Subject: [PATCH] Add TNTPrimeEvent
 
 
 diff --git a/src/main/java/net/minecraft/server/BlockFire.java b/src/main/java/net/minecraft/server/BlockFire.java
-index 48b52b58ff4779fe92d17ab3b3c4cf608917a325..acd0ec4bf30c5d559d99ec449fef166ce59076e6 100644
+index e79f3872de2989d0b9b28a8b49cc399968f5bdbf..4716f366342a3ef70d0ea50e3747786fad1e3573 100644
 --- a/src/main/java/net/minecraft/server/BlockFire.java
 +++ b/src/main/java/net/minecraft/server/BlockFire.java
 @@ -3,6 +3,7 @@ package net.minecraft.server;

--- a/Spigot-Server-Patches/0569-PortalCreateEvent-needs-to-know-its-entity.patch
+++ b/Spigot-Server-Patches/0569-PortalCreateEvent-needs-to-know-its-entity.patch
@@ -22,7 +22,7 @@ index b6081d259444f0d78736802b5d565faf776e825f..505d40278d4472e0015a53af8e93bd81
      public void onPlace(IBlockData iblockdata, World world, BlockPosition blockposition, IBlockData iblockdata1, boolean flag) {
          org.spigotmc.AsyncCatcher.catchOp("block onPlace"); // Spigot
 diff --git a/src/main/java/net/minecraft/server/BlockFire.java b/src/main/java/net/minecraft/server/BlockFire.java
-index 5ef90510b8bdac2dcf58cabfa78ba48a30a22a97..e2dd27be6b523c6e6668923c1b1900cb9251f39f 100644
+index e5d75a8a93ffe7bc3747b48210eea6189a956d54..0e6ce21db67447d0eb84eb1633653decc172b2da 100644
 --- a/src/main/java/net/minecraft/server/BlockFire.java
 +++ b/src/main/java/net/minecraft/server/BlockFire.java
 @@ -343,8 +343,10 @@ public class BlockFire extends BlockFireAbstract {

